File: LeetCode-Python/EvaluateReversePolishNotation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        operators = {'+': lambda x, y: x + y,
                    '-': lambda x, y: x - y,
                    '*': lambda x, y: x * y,
                    '/': lambda x, y: int(x / y)}  # Division truncates toward zero

        for token in tokens:
            if token not in operators:
                stack.append(int(token))
            else:
                operand2 = stack.pop()
                operand1 = stack.pop()
                operation = operators[token]
                result = operation(operand1, operand2)
                stack.append(result)

        return stack[0]

File: LeetCode-Cpp/EvaluateReversePolishNotation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> s;

        for (const string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                int operand2 = s.top(); s.pop();
                int operand1 = s.top(); s.pop();
                if (token == "+") s.push(operand1 + operand2);
                else if (token == "-") s.push(operand1 - operand2);
                else if (token == "*") s.push(operand1 * operand2);
                else if (token == "/") s.push(operand1 / operand2);
            } else {
                s.push(stoi(token));
            }
        }

        return s.top();

    }
};
